pkg/agent/flexvolume/controller.go

```diff
@@ -0,0 +1,341 @@
+		// pod namespace should be the same as the PVC namespace
+		attachOptions.PodNamespace = pv.Spec.ClaimRef.Namespace
+	}
+
+	node := os.Getenv(k8sutil.NodeNameEnvVar)
+	if attachOptions.Pod == "" {
+		// Find all pods scheduled to this node
+		opts := metav1.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector("spec.nodeName", node).String(),
+		}
+		pods, err := c.clientset.Core().Pods(attachOptions.PodNamespace).List(opts)
+		if err != nil {
+			return fmt.Errorf("failed to get pods in namespace %s: %+v", attachOptions.PodNamespace, err)
+		}
+
+		pod := findPodByID(pods, types.UID(attachOptions.PodID))
+		if pod != nil {
+			attachOptions.Pod = pod.GetName()
+		}
+	}
+
+	if attachOptions.Image == "" {
+		attachOptions.Image = pv.Spec.PersistentVolumeSource.FlexVolume.Options[ImageKey]
+	}
+	if attachOptions.Pool == "" {
+		attachOptions.Pool = pv.Spec.PersistentVolumeSource.FlexVolume.Options[PoolKey]
+	}
+	if attachOptions.StorageClass == "" {
+		attachOptions.StorageClass = pv.Spec.PersistentVolumeSource.FlexVolume.Options[StorageClassKey]
+	}
+	attachOptions.ClusterName, err = c.parseClusterName(attachOptions.StorageClass)
+	if err != nil {
+		return fmt.Errorf("Failed to parse clusterName from storageClass %s: %+v", attachOptions.StorageClass, err)
+	}
+	return nil
+}
+
+// GetGlobalMountPath generate the global mount path where the device path is mounted.
+// It is based on the kubelet root dir, which defaults to /var/lib/kubelet
+func (c *FlexvolumeController) GetGlobalMountPath(volumeName string, globalMountPath *string) error {
+	*globalMountPath = path.Join(c.getKubeletRootDir(), "plugins", FlexvolumeVendor, FlexvolumeDriver, "mounts", volumeName)
+	return nil
+}
+
+// getKubeletRootDir queries the kubelet configuration to find the kubelet root dir. Defaults to /var/lib/kubelet
+func (c *FlexvolumeController) getKubeletRootDir() string {
+	nodeConfigURI, err := k8sutil.NodeConfigURI()
+	if err != nil {
+		logger.Warningf(err.Error())
+		return kubeletDefaultRootDir
+	}
+
+	// determining where the path of the kubelet root dir and flexvolume dir on the node
+	nodeConfig, err := c.clientset.Core().RESTClient().Get().RequestURI(nodeConfigURI).DoRaw()
+	if err != nil {
+		logger.Warningf("unable to query node configuration: %v", err)
+		return kubeletDefaultRootDir
+	}
+	configKubelet := agent.NodeConfigKubelet{}
+	if err := json.Unmarshal(nodeConfig, &configKubelet); err != nil {
+		logger.Warningf("unable to parse node config from Kubelet: %+v", err)
+		return kubeletDefaultRootDir
+	}
+
+	if configKubelet.ComponentConfig.RootDirectory == "" {
+		return kubeletDefaultRootDir
+	}
+	return configKubelet.ComponentConfig.RootDirectory
+}
+
+// getPodAndPVNameFromMountDir parses pod information from the mountDir
+func getPodAndPVNameFromMountDir(mountDir string) (string, string, error) {
+	// mountDir is in the form of <rootDir>/pods/<podID>/volumes/rook.io~rook/<pv name>
+	token := strings.Split(mountDir, "/")
```

Comment:
This is given to us by Kubernetes and i dont see other plugins use filepath.Clean() so i am kinda hesitant. Let me read about what filepath.Clean() does first. Is there any concern about the format of mountDir?

Reply:
no i checked the format of mountDir and since we are tokenizing from the end of the path, instead of the beginning, i think the format is OK.  The <rootDir> could be arbitrarily long (contain N number of segments), but since we are counting from the end of the path I think we are OK.